Improve cpu/Loops.h arity asserts.

gchanan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #38817 Add dynamic_cast asserts to CPU Loops.
* #38815 Support void return type in TensorIteratorDynamicCasting checks.
* #38813 Move needs_dynamic_casting to a non-CUDA specific file.
* #38810 Improve error checking of CUDALoops.
* **#38809 Improve cpu/Loops.h arity asserts.**

This splits the asserts into separate input/output asserts and makes the numbers precise, instead of ranges.

This is an ongoing effort to improve the Loops assertion and to integrate dynamic cast checking into CPU loops.

Differential Revision: [D21670263](https://our.internmc.facebook.com/intern/diff/D21670263)

gchanan committed:
Improve cpu/Loops.h arity asserts.

This splits the asserts into separate input/output asserts and makes the numbers precise, instead of ranges.

This is an ongoing effort to improve the Loops assertion and to integrate dynamic cast checking into CPU loops.

[ghstack-poisoned]

gchanan committed:
Update on "Improve cpu/Loops.h arity asserts."


This splits the asserts into separate input/output asserts and makes the numbers precise, instead of ranges.

This is an ongoing effort to improve the Loops assertion and to integrate dynamic cast checking into CPU loops.

Differential Revision: [D21670263](https://our.internmc.facebook.com/intern/diff/D21670263)

[ghstack-poisoned]

gchanan committed:
Update on "Improve cpu/Loops.h arity asserts."


This splits the asserts into separate input/output asserts and makes the numbers precise, instead of ranges.

This is an ongoing effort to improve the Loops assertion and to integrate dynamic cast checking into CPU loops.

Differential Revision: [D21670263](https://our.internmc.facebook.com/intern/diff/D21670263)

[ghstack-poisoned]

ngimel commented on aten/src/ATen/native/cpu/Loops.h:
```diff
@@ -186,7 +186,9 @@ static inline void unroll_contiguous_scalar_checks(
-  TORCH_INTERNAL_ASSERT(iter.ntensors() >= traits::arity + 1);
+  // this could be extended to work with void return types
+  TORCH_INTERNAL_ASSERT(iter.ninputs() == traits::arity);
+  TORCH_INTERNAL_ASSERT(iter.noutputs() == 1);
```
This is true for current kernels, but in general we should not hardcode number of outputs to 1? E.g. divmod will need 2 outputs, some activation backward functions already need 2 outputs, and nvidia has a PR out that enables multiple outputs for gpu tensorIterator kernels https://github.com/pytorch/pytorch/pull/37969/files.

gchanan replied:
thanks for pointing this out, I wasn't aware of that PR.

I'm happy to extend any of these if we need, I just wanted to start with the most restrictive.  So I can change the comments from "// this could be extended to work with void return types" to "// this could be extended to work with void or multiple (tuple) return types"

ngimel commented on aten/src/ATen/native/cpu/Loops.h:
```diff
@@ -226,8 +230,9 @@ void cpu_kernel_vec(TensorIterator& iter, func_t&& op, vec_func_t&& vop) {
-  TORCH_INTERNAL_ASSERT((std::is_void<typename traits::result_type>::value &&
-    iter.noutputs() == 0 && iter.ntensors() == traits::arity) || (iter.ntensors() >= traits::arity + 1));
+  constexpr bool result_void = std::is_void<typename traits::result_type>::value;
+  TORCH_INTERNAL_ASSERT(iter.ninputs() == traits::arity &&
```
This is a different condition than used to be here (previously, result of void type was valid with non-zero noutputs, now it is not). Is this intended?

gchanan replied:
yes, I meant these to be as strict as possible (while still passing the tests).  I'm totally happy to loosen these later if we need to, but they should be based on what we actually need.

gchanan committed:
Update on "Improve cpu/Loops.h arity asserts."


This splits the asserts into separate input/output asserts and makes the numbers precise, instead of ranges.

This is an ongoing effort to improve the Loops assertion and to integrate dynamic cast checking into CPU loops.

Differential Revision: [D21670263](https://our.internmc.facebook.com/intern/diff/D21670263)

[ghstack-poisoned]
